src/main/java/org/opensearch/neuralsearch/processor/NormalizationProcessorWorkflow.java

```diff
@@ -226,6 +252,201 @@ private void updateOriginalQueryResults(final CombineScoresDto combineScoresDTO,
         }
     }
 
+    private int collapse(
+        List<CompoundTopDocs> queryTopDocs,
+        int indexOfFirstNonEmpty,
+        List<QuerySearchResult> querySearchResults,
+        Sort sort,
+        final boolean isFetchPhaseExecuted,
+        CombineScoresDto combineScoresDTO
+    ) {
+        boolean isKeyword = ((CollapseTopFieldDocs) queryTopDocs.get(indexOfFirstNonEmpty)
+            .getTopDocs()
+            .getFirst()).collapseValues[0] instanceof BytesRef;
+
+        if (isKeyword) {
+            return collapseOnKeyword(queryTopDocs, querySearchResults, sort, indexOfFirstNonEmpty, isFetchPhaseExecuted, combineScoresDTO);
+        } else {
+            return collapseOnNumeric(queryTopDocs, querySearchResults, sort, indexOfFirstNonEmpty, isFetchPhaseExecuted, combineScoresDTO);
+        }
+    }
+
+    private int collapseOnKeyword(
+        List<CompoundTopDocs> queryTopDocs,
+        List<QuerySearchResult> querySearchResults,
+        Sort sort,
+        int indexOfFirstNonEmpty,
+        final boolean isFetchPhaseExecuted,
+        CombineScoresDto combineScoresDTO
+    ) {
+        int totalScoreDocsCount = 0;
+        Map<BytesRef, FieldDoc> collapseValueToTopScoreDocMap = new HashMap<>();
+        Map<BytesRef, Integer> collapseValueToShardMap = new HashMap<>();
+        String collapseField = "";
+
+        HybridQueryFieldDocComparator comparator = new HybridQueryFieldDocComparator(
+            ((CollapseTopFieldDocs) queryTopDocs.get(indexOfFirstNonEmpty).getTopDocs().getFirst()).fields,
+            Comparator.comparing((ScoreDoc scoreDoc) -> scoreDoc.score)
+        );
+
+        for (int i = 0; i < querySearchResults.size(); i++) {
+            CompoundTopDocs updatedTopDocs = queryTopDocs.get(i);
+            List<ScoreDoc> updatedScoreDocs = updatedTopDocs.getScoreDocs();
+            if (updatedScoreDocs.isEmpty()) {
+                continue;
+            }
+            collapseField = ((CollapseTopFieldDocs) updatedTopDocs.getTopDocs().getFirst()).field;
+
+            for (ScoreDoc scoreDoc : updatedScoreDocs) {
+                FieldDoc fieldDoc = (FieldDoc) scoreDoc;
+                BytesRef collapseValue = (BytesRef) fieldDoc.fields[fieldDoc.fields.length - 1];
```

Comment:
add validation for fields array that it's not null and has non-zero length. In case any of those is true we probably need to log and continue to next iteration in the loop